Coupled-Cluster/RHF/utils.py

```diff
@@ -0,0 +1,168 @@
+    """
+    inp, output_ind = input_string.split('->')
+    input_left, input_right = inp.split(',')
+
+    size_dict = {}
+    for s, size in zip(input_left, op1.shape):
+        size_dict[s] = size
+    for s, size in zip(input_right, op2.shape):
+        size_dict[s] = size
+
+    set_left = set(input_left)
+    set_right = set(input_right)
+    set_out = set(output_ind)
+
+    idx_removed = (set_left | set_right) - set_out
+    keep_left = set_left - idx_removed
+    keep_right = set_right - idx_removed
+
+    # Tensordot axes
+    left_pos, right_pos = (), ()
+    for s in idx_removed:
+        left_pos += (input_left.find(s),)
+        right_pos += (input_right.find(s),)
+    tdot_axes = (left_pos, right_pos)
+
+    # Get result ordering
+    tdot_result = input_left + input_right
+    for s in idx_removed:
+        tdot_result = tdot_result.replace(s, '')
+
+    rs = len(idx_removed)
+    dim_left, dim_right, dim_removed = 1, 1, 1
+    for key, size in size_dict.items():
+        if key in keep_left:
+            dim_left *= size
+        if key in keep_right:
+            dim_right *= size
+        if key in idx_removed:
+            dim_removed *= size
+
+    shape_result = tuple(size_dict[x] for x in tdot_result)
+    used_einsum = False
+
+    # Matrix multiply
+    # No transpose needed
+    if input_left[-rs:] == input_right[:rs]:
+        new_view = np.dot(op1.reshape(dim_left, dim_removed),
+                          op2.reshape(dim_removed, dim_right))
+
+    # Transpose both
+    elif input_left[:rs] == input_right[-rs:]:
+        new_view = np.dot(op1.reshape(dim_removed, dim_left).T,
+                          op2.reshape(dim_right, dim_removed).T)
+
+    # Transpose right
+    elif input_left[-rs:] == input_right[-rs:]:
+        new_view = np.dot(op1.reshape(dim_left, dim_removed),
+                          op2.reshape(dim_right, dim_removed).T)
+
+    # Tranpose left
+    elif input_left[:rs] == input_right[:rs]:
+        new_view = np.dot(op1.reshape(dim_removed, dim_left).T,
+                          op2.reshape(dim_removed, dim_right))
+
+    # If we have to transpose vector-matrix, einsum is faster
+    elif (len(keep_left) == 0) or (len(keep_right) == 0):
+        new_view = np.einsum(input_string, op1, op2)
+        used_einsum = True
+
+    else:
+        new_view = np.tensordot(op1, op2, axes=tdot_axes)
+
+    # Make sure the resulting shape is correct
+    if (new_view.shape != shape_result) and not used_einsum:
+        if (len(shape_result) > 0):
+            new_view = new_view.reshape(shape_result)
+        else:
+            new_view = np.squeeze(new_view)
+
+    # In-place mult by prefactor if requested
+    if prefactor is not None:
+        new_view *= prefactor
+
+    # Do final tranpose if needed
+    if used_einsum:
+        return new_view
+    elif tdot_result == output_ind:
+        return new_view
+    else:
+        return np.einsum(tdot_result + '->' + output_ind, new_view)
```

Comment:
We might be able to kill off `ndot` in the future as this functionality is now baked into `einsum` itself.

Reply:
Ok. But one needs to have the latest version of numpy.